app/controllers/assessments_controller.rb

```diff
@@ -119,46 +118,35 @@ def score
     authorize @assessment
     # XXX - what else needs to be authorized? Enrollment?
 
-    # binding.pry
-
     if @assessment.due_date > Time.current
-      #
-      # Not due yet, only score published Posts.
-      #
-      posts = Post.where(assessment: params[:id], published: true)
-      strands = posts[0].strands.count
-      post_ids = posts.map &:id
-      posts.map do |p|
-        if p.scores.count == strands
-          # Post appears to have been previously scored.
-          post_ids.delete(p.id)
-        end
-      end
-      @posts = posts.where(id: post_ids)
-      @missing = []
-      # binding.pry
+      # Due date is still in the future, only score published posts.
+      @posts = Post.where(assessment: params[:id], published: true)
+      @missing = nil
     else
-      #
       # Due date has passed, set up to score everything.
-      #
-      students = Enrollment.where(section: @assessment.section)
-      missing = students.map &:student_id
-      posts = Post.where(assessment: params[:id])
-      strands = Assessment.find(params[:id]).strands.count
-      post_ids = posts.map &:id
-      posts.map do |p|
-        if p.scores.count == strands
-          # Post appears to have been previously scored.
-          post_ids.delete(p.id)
-        end
-        missing.delete(p.user_id)
+      @posts = Post.where(assessment: @assessment)
+      @missing = User.where(
+        id: (Enrollment.where(section: @assessment.section).map &:student_id))
+                     .order(family_name: :asc, given_name: :asc)
+    end
+
+    strands = Assessment.find(@assessment).strands.count
+
+    @posts.select do |p|
+      @missing.empty? || @missing -= [User.find(p.user_id)]
+
+      p.scores.count != strands
+    end
+
+    unless @missing.empty?
+      # Remove users with missing posts that have already been scored.
+      scores = Score.where(user: @missing, assessment: @assessment)
+      @missing.select! do |u|
+        scores.where(user: u).count != strands
       end
-      @posts = posts.where(id: post_ids)
-      # TODO: need to check that @missing doesn't contain assessments that have
-      #       already been scored.
-      @missing = User.where(id: missing)
-      # binding.pry
     end
+
+    @posts.includes(:user).order('users.family_name ASC, users.given_name ASC')
```

Comment:
`includes`, `order`, `where`, and friends don't work in place. Options:

1. Replace them with their `!` variants.
2. Merge this with line 127 (initial creation of @posts)
3. `@posts = @posts.includes…`